### ISSUE TITLE
Open multiple instances / Windows

### ISSUE THREAD

**Issue by amitkot** [Kethku/neovide#1332]
**Is your feature request related to a problem? Please describe.**
I cannot have more than one Neovide window running at the same time.
 
**Describe the solution you'd like**
Allow running multiple instances of Neovide.

**Additional context**
I use macOS.

**Comment by yuhuihu** [comment]
I need multiple instances too.

**Comment by AndreyGladkov** [comment]
The procedure to run multiple instances is as follows:
1. Open the Terminal application
2. Enter `neovide`
3. Press Enter

**Comment by JulienLecoq** [comment]
Arg, I hate that macOS lacks of such feature in Spotlight

**Comment by JulienLecoq** [comment]
Still looking for a way to do that with an alternative of Spotlight 😅

**Comment by tonogdlp** [comment]
I use the Alfred App instead of Spotlight, a game changer in many ways... But for you, you can do the following prompt:

`>neovide`

It will run the command after the `>` .. in this case the nevoid command

**Comment by kipelovets** [comment]
It's cool there's a workaround, but the experience is certainly lacking compared to MacVim for instance. MacVim's multiple windows can be cycled through with ⌘+` and closed with ⌘+w, while the whole application is closed with ⌘+q. On the other hand the new instances of Neovide are separate applications that pollute the ⌘+Tab list and have to be closed individually.

**Comment by treeman** [comment]
One use-case this would enable is with multiple monitors you could have a window with a compilation buffer or diagnostics on the secondary monitor while having the main monitor full of code. This would enable you to compile from within neovim, instead of having to jump to a separate terminal, and would keep the diagnostics list.

This is in my view the big differentiator between neovim in a gui vs neovim in a terminal.

**Comment by fredizzimo** [comment]
Neovim actually do support multiple windows that would be completely compatible with Neovide

https://neovim.io/doc/user/api.html#nvim_open_win()
>  external: GUI should display the window as an external
   top-level window. Currently accepts no other positioning
   configuration together with this.

There's also a built-in keymapping for it https://neovim.io/doc/user/windows.html#CTRL-W_ge

We then receive the following UI event
> ["win_external_pos", grid, win]  
  Display or reconfigure external window `win`. The window should be
  displayed as a separate top-level window in the desktop environment,
  or something similar.

I think it's actually a bug to not implement it, it's part of the standard protocol without any extensions. And if you do
1. `:split`
2. `<CTRL-W>e`
3. `i` and type something, Neovim now thinks that the invisible external window is focused, the cursor is not visible for example

Of course these external windows are not full fledged Neovim instances, I'm not sure but I don't think they:
1. Have a cmd line
2. Can be split futher
3. You can open floating windows anchored to them (but this one I'm not sure about at all)

But I think you can:
1. Have a winbar
2. Switch buffers
3. Resize them completely independently from the main window

So at least for people with multiple monitors, it could be useful. And as mentioned we are actually required to implement it according to the specs, so re-opening this.

That said, it's not a small task to implement it, but I don't see any huge obstacles either.

**Comment by Kethku** [comment]
I agree with this. Depending on energy I may take a stab at implementing this especially soonish

**Comment by bqv** [comment]
What would be the steps to implement this?

**Comment by kentslaney** [comment]
In case the executable didn't end up in your `$PATH` (which was the case for me), it's located at `/Applications/neovide.app/Contents/MacOS/neovide`

**Comment by falcucci** [comment]
for multiple instance management we also have Neohub available for macOS.

Use this release meanwhile.

https://github.com/falcucci/NeoHub/releases/tag/v0.0.2

**Comment by iwinux** [comment]
Here's my current workaround with Hammerspoon:

<img width="1020" alt="image" src="https://github.com/neovide/neovide/assets/98864/27947020-3fd1-484e-a3f5-78500c7659d8">

```lua
function cycleNeovideWindows()
  local current = hs.window.focusedWindow()

  local items = hs.fnutils.imap({hs.application.find('com.neovide.neovide')}, function(app)
    local title = app:title()
    local status
    local win = app:mainWindow()

    if win ~= nil then
      title = win:title()
    end

    if win == current then
      status = '[CURRENT]'
    end

    return {
      text = title,
      subText = status,
      pid = app:pid(),
    }
  end)

  local callback = function(result)
    hs.application.applicationForPID(result.pid):activate()
  end

  hs.chooser.new(callback):choices(items):show()
end

hs.hotkey.bind({'cmd', 'ctrl'}, '`', cycleNeovideWindows)
```

Press <kbd>Command + Ctrl + `</kbd> to choose Neovide instances. **NOTE:** Neovide instances need to be launched with `open --new -b com.neovide.neovide`, otherwise they can't be looked up with bundle ID `com.neovide.neovide`.

**Comment by samgj18** [comment]
@iwinux I did use your script. However, I found particularly useful to modify the command as follows:

```lua
open --new -b com.neovide.neovide --args ${PWD}
```

This is similar to do ".", which I use pretty often.

**Comment by kentslaney** [comment]
@samgj18 oh cool, that's better than my hard linked path solution

**Comment by preland** [comment]
Ok, so weird question: Is this already implemented somehow?

Context: A couple of days ago I was messing around with keybinds, and accidentally moved one of my tabs into a new window. I don't know how I did it, and I can't remember how I did it. 

(Platform: Hyprland (Arch Linux), NvChad config)

**Comment by davibe** [comment]
Vimr does this right. It opens multiple windows and each windows title matches the titlestring also on macos. Unfortunately it's not as stable as neovide is in my case, it often crashes.

**Comment by bassamsdata** [comment]
I’ve been running intothe issue with the Aerospace tiling window manager where Neovide doesn’t get recognized if it’s opened from the terminal, appearing without a bundle ID in the `aerospace list-apps` output:

```sh
8668 | NULL-APP-BUNDLE-ID | neovide
```

Launching Neovide with the command below works:

```sh
open --new -b com.neovide.neovide --args ${PWD}
```

However, this defaults to the home directory as the working directory, which is why I prefer opening it from the terminal in the first place.

If anyone has suggestions for a workaround, I’d be grateful!

**Comment by bassamsdata** [comment]
@ChristinWhite Thank you for the info and solution.

I did that earlier, and it worked with this configuration:
```toml
[[on-window-detected]]
if.window-title-regex-substring = 'neovide'
run = ['layout tiling', "move-node-to-workspace 3_neovide"]  
```

However, I encountered an issue where Neovide wouldn't focus after moving to the workspace. Even with focus commands like `focus left` or `focus right`, the mouse would center on Neovide, but the app itself wouldn’t get focused. I'm not sure if you've had a similar experience.

```toml
alt-n = 'workspace 3_neovide'
alt-h = 'focus left'
alt-k = 'focus up'
```

To resolve this, I followed a suggestion from this [comment](https://github.com/nikitabobko/AeroSpace/issues/575#issuecomment-2445528574) and created a wrapper file named `neovide-wrapper`. I placed this file inside `/Applications/Neovide.app/Contents/MacOS` and updated the `CFBundleExecutable` in `/Applications/Neovide.app/Contents/info.plist` to point to the wrapper instead. Here's the wrapper:

```sh
#!/bin/bash
# Get the current working directory
CWD="$PWD"

# Get the path to the actual neovide binary in the same directory as this wrapper
NEOVIDE_PATH="$(dirname "$0")/neovide"

# Change to the directory where the command was invoked
cd "$CWD"

# Execute neovide with any passed arguments
"$NEOVIDE_PATH" "$@" &
disown
```

It’s a bit of a workaround, but it solved all my issues. Now, Neovide is fully recognized by Aerospace and actually recognized the working directory where the command was invoked.

**Comment by Sebas-h** [comment]
I got a workaround working to be able to `` Cmd+` `` through the windows (as if they belonged to the same application instance)
Unfortunately, it will still clog up the dock with Neovide application icons.

It uses KarabinerElements and Hammerspoon to accomplish this, as follows:

* KarabinerElements: map `` Cmd+` `` to `` Cmd+Ctrl+Shift+Option+` `` (i.e. an unlikely key combo to already be in use) if the front-most/active application is Neovide
* Hammerspoon: map a function to `` Cmd+Ctrl+Shift+Option+` `` that cycles through all Neovide application instances' windows (requires a bit of state management as well, as the Hammerspoon code below shows)

<details>
<summary>KarabinerElements code</summary>

```json
{
    "description": "If app=neovide then change left_command+` to hyper+`",
    "manipulators": [
        {
            "conditions": [
                {
                    "bundle_identifiers": [
                        "^.*neovide.*$"
                    ],
                    "type": "frontmost_application_if"
                }
            ],
            "from": {
                "key_code": "grave_accent_and_tilde",
                "modifiers": { "mandatory": ["left_command"] }
            },
            "to": [
                {
                    "key_code": "grave_accent_and_tilde",
                    "modifiers": ["left_command", "left_option", "left_shift", "left_control"]
                }
            ],
            "type": "basic"
        }
    ]
}
```

</details>

<details>
<summary>Hammerspoon code</summary>

```lua
-- Set up state
local state = {
    matchingWindows = {},
    currentIndex = 1,
    currentlyCyclingWindows = false,
}

-- Helper functions
local function tableContains(tableT, value)
    for i = 1, #tableT do
        if tableT[i] == value then
            return true
        end
    end
    return false
end

local function repopulateMatchingWindows(appPID, focusedWindowID)
    local firstWinPID = 0 -- Assuming no process ever has PID=0
    local attemptCounter = 0
    local attemptLimit = 10 -- In practice it seems we don't ever get past 2 attempts

    while attemptCounter < attemptLimit and firstWinPID ~= appPID do
        attemptCounter = attemptCounter + 1
        -- Get ordered list of windows on screen (front most window first)
        local orderedWindows = hs.window.orderedWindows()
        if not orderedWindows then
            return
        end

        -- (Re-)populate `matchingWindows`
        state.matchingWindows = {}
        for _, win in ipairs(orderedWindows) do
            if win:application():name() == "Neovide" then
                table.insert(state.matchingWindows, win)
            end
        end

        firstWinPID = 0
        if #state.matchingWindows > 1 then
            firstWinPID = state.matchingWindows[1]:application():pid()
        end
    end

    if state.matchingWindows[1]:application():pid() == appPID then
        state.currentIndex = 1
    else
        -- NOTE: Ideally we shouldn't get here bc it means ordering
        -- returned from `orderedWindows()` was incorrect event after
        -- several (attemptLimit) attempts at retrieving it.
        -- We're setting the currentIndex here as a fallback (so things
        -- still work, in some way)
        print("[DEBUG] Num attempts:", attemptCounter)
        for i, win in ipairs(state.matchingWindows) do
            if win:id() == focusedWindowID then
                state.currentIndex = i
                break
            end
        end
    end
end

-- Application watcher (making this `local` seems to degrade performance)
neovideApplicationWatcher = hs.application.watcher.new(function(appNameStr, eventType, app)
    -- Event type values see: https://github.com/Hammerspoon/hammerspoon/blob/cdd19a5def652540af23e0219acfce5296c3ae7a/extensions/application/libapplication_watcher.m#L54
    if appNameStr ~= "Neovide" then
        return
    end
    if eventType ~= 1 and eventType ~= 5 then
        return
    end
    if not state.currentlyCyclingWindows then
        -- NOTE: For evenType=1 (opening a new window) it seems to take a while
        -- for the window to actually show up so we loop for a few seconds here
        -- to wait for it to arrive (seems reasonable to give up waiting if
        -- opening a new window takes longer than 5 seconds)
        local attemptCounter = 0
        local attemptLimit = 10
        local focusedWindow = app:focusedWindow()
        while attemptCounter < attemptLimit and not focusedWindow do
            attemptCounter = attemptCounter + 1
            hs.timer.usleep(500000) -- 0.5sec sleep
            focusedWindow = app:focusedWindow()
        end
        repopulateMatchingWindows(app:pid(), focusedWindow:id())
    end
end)

-- Function to handle the custom behavior on key combo
function handleHotkeyNeovideSwitcher()
    local focusedApp = hs.application.frontmostApplication()
    if focusedApp and focusedApp:name() == "Neovide" then
        -- Nothing to cycle through bc there's only one Neovide app window
        if #state.matchingWindows < 2 then
            return
        end

        state.currentlyCyclingWindows = true

        -- Move to the next window
        local nextIndex = (state.currentIndex % #state.matchingWindows) + 1
        state.matchingWindows[nextIndex]:focus()
        state.currentIndex = nextIndex

        -- Start listening for Command-key keyUp event we use this as a
        -- proxy/surrogate for: "we stopped cycling throught the app windows"
        commandKeyTap:start()
    end
end

-- Create an event tap to monitor keyboard events
commandKeyTap = hs.eventtap.new({ hs.eventtap.event.types.flagsChanged }, function(event)
    -- Check if the Cmd-key was released
    local flags = event:getFlags()
    if not flags.cmd and event:getKeyCode() == hs.keycodes.map.cmd then
        state.currentlyCyclingWindows = false
        commandKeyTap:stop()
    end
    return false -- Do not block the event
end)

--------------------------------------------------------------------------------
--------------------------------------------------------------------------------

-- Bind the hotkey
hs.hotkey.bind({ "shift", "ctrl", "alt", "cmd" }, "`", handleHotkeyNeovideSwitcher)

-- Start Neovide application watcher
neovideApplicationWatcher:start()
```

</details>

For a less state-heavy, more easily understood but alas slower Hammerspoon version see below

<details>
<summary>Fewer LOC but slower Hammerspoon code</summary>

```lua
local hotkey = {"cmd", "shift", "ctrl", "alt", "`"}
local neovide = "Neovide"

local function handleHotkey()
    local focusedApp = hs.application.frontmostApplication()
    if focusedApp and focusedApp:name() == "Neovide" then
        local focusedWindow = hs.window.focusedWindow()
        if not focusedWindow then return end

        -- Find all windows of applications with the same name
        local allWindows = hs.window.allWindows()
        local matchingWindows = {}

        for _, win in ipairs(allWindows) do
            if win:application():name() == "Neovide" then
                table.insert(matchingWindows, win)
            end
        end

        -- Check if any other windows to cycle!
        if #matchingWindows < 2 then return end

        -- Find the index of the currently focused window
        local currentIndex = nil
        for i, win in ipairs(matchingWindows) do
            if win:id() == focusedWindow:id() then
                currentIndex = i
                break
            end
        end

        -- Move to the next window
        if currentIndex then
            local nextIndex = (currentIndex % #matchingWindows) + 1
            matchingWindows[nextIndex]:focus()
        end
    end
end

-- Bind the function to a hotkey
hs.hotkey.bind({'shift', 'ctrl', 'alt', 'cmd'}, '`', handleHotkey)
```

</details>
---

(I only have a single monitor, so I'm not sure this works as expected on a multi-monitor setup)
(I also use `:silent exec '!open --new -b com.neovide.neovide'<cr>` (as mentioned in this issue) as the nvim keymap to open a new window, seems to work fine as of v0.14.0)

**Comment by falcucci** [comment]
Just to keep everyone in track of it, this is supposed to be working very soon in the multi-window refactor I am doing at https://github.com/neovide/neovide/pull/2872 pull request. As long as I am arriving at some stability of the core changes, the many Neovide icons (instances) visible in your dock will not be a problem anymore and this should work as only one instance for macOS.

Updates coming soon.

**Comment by erlangparasu** [comment]
any updates?

**Comment by theol0403** [comment]
> if vim.g.neovide and vim.fn.has('macunix') then
>     vim.keymap.set("n", "<D-n>", ":silent exec '!/Applications/neovide.app/Contents/MacOS/neovide'<cr>")
> end

I found that this blocks the original window until the spawned window is closed. This fixes the issue for me:

```lua
  vim.keymap.set("n", "<D-n>", function()
    vim.fn.jobstart('cd ~ && neovide --fork')
  end, {silent = true})
```

**Comment by iwinux** [comment]
Honestly I have switch to [VimR](https://github.com/qvacua/vimr) since my last comment and it works great :)

**Comment by farcaller** [comment]
VimR has a huge downside of bundling its own neovim, which makes it impossible to use if your neovim is customized in any way.

**Comment by unphased** [comment]
Probably worth trying to do a small survey. This is for my own notes but I'm doing it publicly:

1. #2872 code is missing, hopefully comes back some time, but anyway my understanding of "multi-window support" would be to have a single neovide app instance support macos native style multiple windows in app behavior: If you have 3 neovide windows open (one instance of neovide.app), 3 instances of neovim exist, backing each of these windows. IMO this is the holy grail for MacOS as it fits perfectly with the way native Mac apps work. On other OS's the window manager has no ".app" notion associating windows of a given app together. 
2. NeovideProject -- i tested this and it did not work. 
3. My current usage pattern: I tend to open a neovide from zsh shell in my projects via `( neovide & )`. This gives it the correct cwd on startup and the neovim level plugin i use (`'rmagatti/auto-session'`) to manage projects/sessions preserves the state of neovim for me in each given git project. I use this in conjunction with a simple "open or switch to specific app" bettertouchtool keyboard hotkey. The main drawback of this is that if i have multiple neovides open, e.g. multiple code projects open with neovide, then when i hit the hotkey that focuses neovide, it only opens the first one, it is not capable of cycling between the neovides (which (1)'s os native windowing support would provide via native cmd+backtick)
4. Karabiner/Hammerspoon trick via @Sebas-h i think is pretty clever and should give a functional workaround that approaches (1)'s workflow. I'm going to try it next. Main drawback is the application list is polluted by neovide, but it is what it is.
5. Neohub - it works but I just don't really like the idea of having a whole app just to fiddle with instances of neovide and does not reduce app list pollution
6. I had a custom hammerspoon only system for managing neovide instances. I built something around `hs.chooser` and I was doing things like enumerating os windows to find which are neovides so i can render the chooser correctly (which items in there are already currently open, and so on). This worked but as I added features to it hammerspoon began to be sluggish in ways impossible to troubleshoot, so at the time I abandoned it in favor of neohub. It largely works like neohub: open it with a hotkey and you can fuzzy type a project path to quickly open it up in neovide.
7. using the neovim level project picker from `auto-session`, `:SessionSearch` bound to a key works well to take a neovide and neovim instance and switch it to another project. Again the UX here is identical to neohub, but it works at the neovim level so neovide is uninvolved. At the end of the day this is an important baseline ux to have, because I sometimes work on my base model Mac Mini and 16GB ram does not allow you to open very many projects with their LSP servers before you have used up the RAM on the computer, and I need plenty for my browser tabs! 

- I also established something I'm experimenting with now which is to assign each of my projects a color and to have all zsh shells (moderated by tmux) and vims use that color as the bgcolor of the terminal pane or window. This helps me avoid editing the wrong files in the wrong project folders. Sometimes we do insane things with git by cloning a second repo, but lately i also have taken to having AI driven productivity happen in parallel across multiple instances of the same project repo, so the background color is a visual cue that i am getting good mileage out of, and it helps especially since tmux's terminal panes are physically separated from the neovide windows.
- I'm hoping to augment terminal and tmux workflows by building terminal functionality into a web app with a terminal emulator server/client system (sitting transparently at the shell layer) so i can do my work from mobile touch devices, which is helped a lot by a recent new trend of not really needing a keyboard as much since a majority of work is being done through prompts, though tmux is probably going to continue to be my daily driver, being able to branch out to zellij or not use tmux at all would be a nice bit of freedom.

**Comment by falcucci** [comment]
It was very unfortunate that I had decided to remove my contributions in general including the multi-window feature due to internal issues about the neovide governance, which is something that doesn't worth any exposition or explanation to respect everyone's privacy.

That said i'll recompose and update the #2872 from the latest changes and leave it in a useful state when you can fully manage your instances using the neovide 0.15.1, from there on, ill try to update it here and there but *I need to add more features and fix some possible regressions, use it at your own risk. No promises.*

AT this time I can't fully focus on this since I am working on other projects but will try to be as much responsive as possible when I get free time which I literally need to concatenate w/ family but at least this would probably unblock people and give them a more organized way of doing things on macOS.

**Comment by theherk** [comment]
> 2\. NeovideProject -- i tested this and it did not work.

@unphased, I understand this was around the same time that you discovered maybe you were not clear on the usage, but I'd really love to follow up on this. If you could open a bug on [that project](https://github.com/theherk/NeovideProject/issues) maybe we could make it work better for people. A bummer if people just discover a comment that it doesn't work without further context. I use it everyday, and it does a nice job of managing many projects, each with their own socket. That said, it does pollute the app list, but I don't see a way around that until this feature discussed here is complete.

@falcucci, thank you for following up. Bummer that there are struggles due to governance, because I think this is such a wonderful project, and this feature is going to take it to new heights. This will be a major boon for Neovide, and I appreciate all your effort on it.

**Comment by unphased** [comment]
@theherk sorry for the delay, but I would like to help with that. It is clear I did not have a proper understanding of how NeovideProject is intended to work. I reviewed the supplementary video you made and it makes more sense now. I have in the meantime also experimented with building my own url scheme and i realize in order to get a lot of the native macos integrations with interactions you have to create an app to use as a handle with automations and things like that. I'm not sure if this is even the idea behind it, but it does seem to fit that mold to me now as well.

Could you clarify some more things that aren't clear to me from the readme?

- when you say instance it isn't clear to me if it is a neovim or neovide instance. I am under the impression that a single neovide could in theory connect to more than one underlying neovim instance though I don't know if switching that regularly is how it's intended to be used. in my experience large RAM usage from neovide usage come overwhelmingly from LSPs loaded by neovim instances, so that helps us reason about which instances of what we would like to keep minimal. In this case it wouldn't help but I do like being able to have multiple instances if i am switching back and forth a lot. 
- when neovideproject says "Single instance per Git repository" it's not really clear what this means. If you open a dir that is a git repo it will open an instance of neovide there. Then, if i open a file within that repo it should open that as a buffer or tab in neovide and that is sensible but if we call open on it at a subdirectory of this one, what will it do? just focus the existing neovide instance and open nothing? open netrw? (My preference just happens to be for focusing the instance and doing nothing!)
- What things specific to directories and git repos and neovide does NeovideProject actually do? It is important because i use a sessions plugin that will auto restore a session when i open neovide at a given directory, which is something i'm pretty used to and need now. So it makes me worried if this thing has overlapping functionality that may clash. My hope is that it only manages instance separation across projects with git repo serving as sentinels for project boundaries, so i can continue to use nvim plugins to implement auto workspace restore features.  
- the relevance of other tools like zoxide, raycast, fzf are just examples of good UXes around quickly opening up a given project or file, but are not dependencies of the tool. 

At the end of the day i can see it as something I want to use, but it's slightly awkward as it gives me like only a fraction of the functionality I need. 

Which is
1. ability to pull up a fuzzy project lister that launches or focuses a neovide for the project that is chosen
2. ability to cycle to the next open neovide instance (hopefully soon cmd+tilde can address this for neovide)
3. ability to instantly pull up the most recently used neovide instance (I am also lacking this, currently using BTT to focus neovide, but it opens the first, not the most recent.)

NeovideProject can provide 50% of (1), while NeoHub is 100% of (1). The big pro for NeovideProject though is that it seems to be a proper unix "do one job" approach of this, so if I already heavily use raycast or zoxide or something it'd be great. But I don't. I might prefer to make a fzf thing but it's also semi-silly to be launching a whole new terminal emulator instance to run fzf just to get a picker. But fzf is just so good that it may be worth doing. 

Right now my workflow is run neovide in tmux half the time to colocate the UI with the projects' terminals and open up neovide instances for when i actually dive into code. Even though writing code is greatly reduced in the last year, reading code happens just as much if not more, and neovide being smooth helps for reading. I can't open too many neovides though because it's awkward to cycle through them, plus sometimes memory consumption is a problem (but I'm gonna ignore that by pretending the LSP server auto suspension plugin i'm using works reliably, which it doesn't right now).

I might bring back NeoHub to address (1) but it's hollow because of (2) and (3) being missing. I can hack (2) pretty easily with hammerspoon, and preferably something better than hammerspoon, but (3) might actually require hammerspoon. 

Another thought is that I can use lua code in neovim to detect activity and provide both (3) and (2) driven by NeovideProject (pending clarification on some of my questions from up above). For example, I could make two keybinds in BTT that trigger scripts that read state about which neovim's i interacted with recently and delegate to NeovideProject to handle Neovide app focus state. This could be really interesting because it lets me do this weird specific stuff in neovim's own config, and it lets me do it without relying on hammerspoon, and I can also independently modularly add a fuzzy project launch picker that I'll have more customizability with. 

So it's really been a roller coaster for me as to the relevance of this thing!

**Comment by unphased** [comment]
I was able to find a much better BetterTouchTool based flow. Behold: 

<img width="729" height="670" alt="Image" src="https://github.com/user-attachments/assets/d5ef4409-1fd1-4166-ab0e-3ca4fdd4f47b" />

All of my needs 1, 2, and 3 are being addressed. It defaults to the most recent (so, handles 3), i can repeatedly hit the keybind (which is opt+n for me) to achieve 2, and it comes with 1. With the way this works I can add more text to fuzzy match against simply by further fleshing out my neovim `titlestring`. Spectacular... 

The blueprint for this is detailed here: https://community.folivora.ai/t/open-a-specific-app-but-be-able-to-cycle-through-them/45454/4

**Comment by theherk** [comment]
@unphased: In order not to fill up this discussion further, I have added [Detailed Behavior](https://github.com/theherk/NeovideProject/tree/main?tab=readme-ov-file#detailed-behavior) to the documentation where I hope to address your questions.

I try my best to clarify, but it is likely that even after doing so, the intent of NeovideProject will not suit your needs or your desired workflow. The best place to understand what is really intended to happen is in [src/Contents/Resources/neovide-project](https://github.com/theherk/NeovideProject/blob/main/src/Contents/Resources/neovide-project). It is really quite rudimentary but bespoke.

**Comment by unphased** [comment]
@theherk thanks. This document seems overly tailored to my specific questions rather than a concise explanation of what actions it does, but after reading through it, what it does has become clarified, so I appreciate that. 

In terms of my workflow I think it has a place. But, I may have to agree that it may need modification to be an ideal fit *for me* in terms of decisionmaking around when to make new instances vs. when to focus an existing one. I often have instances of neovim open in terminals under tmux and i have independent neovides open. I'll have to test and see how it behaves if i have a project already open in a terminal what NeovideProject will do when invoked on that dir; it may try to do a neovide-focus event on that neovim, which will do nothing, when what I'd expect by launching that is to look to confirm the existing instance is not a neovide-instance of neovim and therefore still launch a neovide there. 

Aside from that one wrinkle though (and some weird issue with my plugins and config that causes a line to be deleted any time a file is opened with a split) this one basically does what I'd want!

**Comment by falcucci** [comment]
I have great news for the multi-window feature refactoring that comes ahead. Personally I love [Neohub](https://github.com/alex35mil/NeoHub), but having an extra app just to handle instances is overkill to me but the Neovide mess at the dock when running multiple instances turns it reasonable to use.

From now on, [this pull request](https://github.com/neovide/neovide/pull/2872) build ships with a pair of system-wide shortcuts that gives the multi-window workflows the same muscle memory that NeoHub users enjoy (including myself). This is going to be splited up to and be available at nighly releases as soon as I create the small stacked pull requests to make the review processes easier but this branch should be the final version to users test and troubleshoot.

A `pinned` shortcut concept (`⌘⌃Z` by default) toggles the most recent used Neovide window: hit it once from anywhere and that editor jumps to the front; hit it again while it's focused and it hides, keeping your desktop tidy. Under the hood we track a most-recently-used list, so you always land on the window you touched last perfectly for bouncing between a terminal and your main Neovide session without touching the mouse.

The second shortcut (`⌘⌃N` by default) opens macOS builtin tab overview, which we rebrand as **Editors** to avoid confusion with **neovim tabs**. If you have more than one Neovide window, this shortcut first activates the latest one (so the overview immediately has focus into that window) and then triggers the system's tab view switcher, letting you picking up any other window with the standard macOS tab-navigation keys `⌘K`/`⌘⇧K` (some must be different but the shortcut is there)

If there's only one window to choose from, we simply fall back to the pinned behavior, so you never get stuck staring at an empty overview. behind the scenes, both shortcuts register through carbon's RegisterEventHotKey, which means the keypress never leaks into whatever app is in the frontier.

The hotkeys are also fully configurable. for now just set `NEOVIDE_MACOS_PINNED_HOTKEY` for the pinned and `NEOVIDE_MACOS_SWITCHER_HOTKEY` for the switcher before launching Neovide to swap in your own combos or set them to `off` / `none`  or empty to disable a shortcut entirely.

If a shortcut does nothing, always double-check `Settings -> Privacy & Security -> Input Monitoring` to ensure the exact Neovide binary or the bundle app itself you're running is listed and checked; macOS doesn't deliver global hotkeys without that permission. double check the env vars or set them to off and relaunch to confirm custom combos or disabled shortcuts behave exactly as you expect.

Internally, there are good improvements done at https://github.com/neovide/neovide/pull/3252 which has already a lot of similatiries applied here, so we will split it up and merge what is similar first. then step by step include on this pull request what is co-related with.

Note: I'll be constantly updating the pull request so be aware.

**Comment by daviareias** [comment]
> I have great news for the multi-window feature refactoring that comes ahead. Personally I love [Neohub](https://github.com/alex35mil/NeoHub), but having an extra app just to handle instances is overkill to me but the Neovide mess at the dock when running multiple instances turns it reasonable to use.

Thank you so much, neovide is by far the terminal with the best readability for me, but not having tabs is the only reason I don't use it at home

**Comment by kgfly** [comment]
This is a great feature. But can multi windows on Linux/Windows support be added to?
